app/frontend/wallet/helpers/CachedDeriveXpubFactory.ts

```diff
@@ -28,14 +37,55 @@ function CachedDeriveXpubFactory(derivationScheme, deriveXpubHardenedFn) {
     return await derivedXpubs[memoKey]
   }
 
-  async function deriveXpubNonhardenedFn(derivationPath) {
+  async function deriveXpubNonhardenedFn(derivationPath: BIP32Path) {
     const lastIndex = derivationPath.slice(-1)[0]
     const parentXpub = await deriveXpub(derivationPath.slice(0, -1))
-
     return deriveChildXpub(parentXpub, lastIndex, derivationScheme.ed25519Mode)
   }
 
-  return deriveXpub
+  function createPathBulk(derivationPath: BIP32Path): BIP32Path[] {
+    const paths: BIP32Path[] = []
+    const accountIndex = derivationPath[2] - HARDENED_THRESHOLD
+    const currentAccountPage = Math.floor(accountIndex / MAX_BULK_EXPORT_AMOUNT)
+
+    /*
+    * in case of the account 0 we append also the byron path
+    * since during byron era only the first account was used
+    */
+    if (accountIndex === 0) paths.push(BYRON_V2_PATH)
+
+    for (let i = 0; i < MAX_BULK_EXPORT_AMOUNT; i += 1) {
+      const nextAccountIndex = currentAccountPage * MAX_BULK_EXPORT_AMOUNT + i + HARDENED_THRESHOLD
+      const nextAccountPath = [...derivationPath.slice(0, -1), nextAccountIndex]
+      paths.push(nextAccountPath)
+    }
+    return paths
+  }
+
+  async function deriveXpubHardenedFn(derivationPath: BIP32Path): Promise<any> {
```

Comment:
deriveXpubFn should imho be renamed back to `deriveXpubHardenedFn` and this function could be then named `_deriveXpubHardenedFn` with a comment above explaining that on top of the original `deriveXpubHardenedFn` this is priming the cache of derived keys to minimize the number of prompts on hardware wallets

Reply:
Why do you think the function that is passed from the cryptoProvider should have the `hardened` keyword in it. It can derive any key not only hardened compared to the `deriveXpubHardenedFn` in `CachedDeriveXpubFactory` which derives only hardened paths. So it makes more sense to to just call the one that is passed from crypto provider `deriveXpubs` and call this one `deriveXpubsHardenedFn`. What do you think? 

(sorry previously I forgot the `s` in pub`s`) so I'm sure it made it more confusing.

Reply:
technically, the function passed from outside is being used only for hardened derivation even if it supported non-hardened derivation. And the non-hardened derivation is performed with the cardano-crypto.js function for that, so even though the function passed from outside would be able to do it, it won't be used for that purpose - although it's rather a low-level concern, the point of this cached derivation is exactly that - that it minimizes the calls to "hardened" derivations which in turn minimizes calls to hw wallets, if used, so from that point of view it's IMHO fair to make it obvious to the reader and not make them misleadingly believe that whatever they pass would be used for non-hardened derivation.